#### These steps are all part of the data preprocessing phase in data analysis and machine learning
- 1. Importing and Loading Data

#### Initial Exploration
- 2. View Basic Data:
- Use methods like df.head(), df.tail(), df.info(), and df.describe() to understand the structure and basic statistics of your dataset.

In [148]:
#Loadig the data set
import pandas as pd # type: ignore
import numpy as np
dataset='Thyroid_Diff.csv'
patient=pd.read_csv(dataset)
print(patient.to_string())

     Age Gender Smoking Hx Smoking Hx Radiothreapy             Thyroid Function         Physical Examination Adenopathy       Pathology     Focality          Risk    T    N   M Stage                Response Recurred
0     27      F      No         No              No                    Euthyroid   Single nodular goiter-left         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I           Indeterminate       No
1     34      F      No        Yes              No                    Euthyroid          Multinodular goiter         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I               Excellent       No
2     30      F      No         No              No                    Euthyroid  Single nodular goiter-right         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I               Excellent       No
3     62      F      No         No              No                    Euthyroid  Single nodular goiter-right         No  Micropapillary 

In [149]:



#Adding a new Column in the dataframe 
patient['patient_ID']=100
print(patient.to_string())




     Age Gender Smoking Hx Smoking Hx Radiothreapy             Thyroid Function         Physical Examination Adenopathy       Pathology     Focality          Risk    T    N   M Stage                Response Recurred  patient_ID
0     27      F      No         No              No                    Euthyroid   Single nodular goiter-left         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I           Indeterminate       No         100
1     34      F      No        Yes              No                    Euthyroid          Multinodular goiter         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I               Excellent       No         100
2     30      F      No         No              No                    Euthyroid  Single nodular goiter-right         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I               Excellent       No         100
3     62      F      No         No              No                    Euthyroid  Single 

In [150]:
print(len(patient.columns))  #before setting Patient_ID as index column

18


In [151]:
List=[]
for id in range(2340,2340+200*len(patient),200):  
        List.append(id)       
patient['patient_ID']=List
#print(patient.to_string())
patient=patient.set_index('patient_ID')
                                                                 



In [152]:
print(len(patient.columns))

17


In [153]:
print(patient.to_string())

            Age Gender Smoking Hx Smoking Hx Radiothreapy             Thyroid Function         Physical Examination Adenopathy       Pathology     Focality          Risk    T    N   M Stage                Response Recurred
patient_ID                                                                                                                                                                                                                    
2340         27      F      No         No              No                    Euthyroid   Single nodular goiter-left         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I           Indeterminate       No
2540         34      F      No        Yes              No                    Euthyroid          Multinodular goiter         No  Micropapillary    Uni-Focal           Low  T1a   N0  M0     I               Excellent       No
2740         30      F      No         No              No                    Euthyroid  Single nodular goite

In [154]:

en_25=patient.head(25)
print(en_25.to_string())

last_10=patient.tail(10)
print(last_10.to_string())

            Age Gender Smoking Hx Smoking Hx Radiothreapy          Thyroid Function         Physical Examination Adenopathy       Pathology     Focality Risk    T   N   M Stage       Response Recurred
patient_ID                                                                                                                                                                                              
2340         27      F      No         No              No                 Euthyroid   Single nodular goiter-left         No  Micropapillary    Uni-Focal  Low  T1a  N0  M0     I  Indeterminate       No
2540         34      F      No        Yes              No                 Euthyroid          Multinodular goiter         No  Micropapillary    Uni-Focal  Low  T1a  N0  M0     I      Excellent       No
2740         30      F      No         No              No                 Euthyroid  Single nodular goiter-right         No  Micropapillary    Uni-Focal  Low  T1a  N0  M0     I      Excellent     

In [155]:


no_of_obs1=len(patient.index)                    #len(patient)  gives same result
no_of_obs2=patient.shape[0]
no_of_obs3=patient.count().max()
print("No of observations are: ",no_of_obs1)
print()
print("No of observations are: ",no_of_obs2)
print()
print(patient.count())
print()
print()
print("No of observations are: ",no_of_obs3)

No of observations are:  383

No of observations are:  383

Age                     383
Gender                  383
Smoking                 383
Hx Smoking              383
Hx Radiothreapy         383
Thyroid Function        383
Physical Examination    383
Adenopathy              383
Pathology               383
Focality                383
Risk                    383
T                       383
N                       383
M                       383
Stage                   383
Response                383
Recurred                383
dtype: int64


No of observations are:  383


In [156]:

no_of_col1=patient.shape[1]
no_of_col2=patient.columns.size                 #patient.columns returns an Index object that contains the column labels. 
no_of_col3=len(patient.columns)                 #The Index object does not have a count() method. 
no_of_col_row=patient.shape                     #The count() method is available for Series and DataFrames but not for Index objects.
print("The no of columns are: ",no_of_col1)
print()
print("The no of columns are: ",no_of_col2)
print()
print("The no of columns are: ",no_of_col3)
print()
print("The no of columns are: ",no_of_col_row)


The no of columns are:  17

The no of columns are:  17

The no of columns are:  17

The no of columns are:  (383, 17)


In [157]:


col_name1=patient.columns
col_name2=patient.columns.to_list()
print("The Column names are :\n",col_name1)
print()
print("The Column names are :\n",col_name2)



The Column names are :
 Index(['Age', 'Gender', 'Smoking', 'Hx Smoking', 'Hx Radiothreapy',
       'Thyroid Function', 'Physical Examination', 'Adenopathy', 'Pathology',
       'Focality', 'Risk', 'T', 'N', 'M', 'Stage', 'Response', 'Recurred'],
      dtype='object')

The Column names are :
 ['Age', 'Gender', 'Smoking', 'Hx Smoking', 'Hx Radiothreapy', 'Thyroid Function', 'Physical Examination', 'Adenopathy', 'Pathology', 'Focality', 'Risk', 'T', 'N', 'M', 'Stage', 'Response', 'Recurred']


In [158]:


row_idx1=patient.index
row_idx2=patient.index.to_list()
col_idx1=patient.columns
col_idx2=patient.columns.to_list()

print("Row indexing is:",row_idx1)
print()
print("Row indexing is:",row_idx2)
print()
print("Column indexing is:",col_idx1)
print()
print("Column indexing is:",col_idx2)
print()

data_type=patient.dtypes
print()
print("Data type of each column is:\n",data_type)


Row indexing is: Index([ 2340,  2540,  2740,  2940,  3140,  3340,  3540,  3740,  3940,  4140,
       ...
       76940, 77140, 77340, 77540, 77740, 77940, 78140, 78340, 78540, 78740],
      dtype='int64', name='patient_ID', length=383)

Row indexing is: [2340, 2540, 2740, 2940, 3140, 3340, 3540, 3740, 3940, 4140, 4340, 4540, 4740, 4940, 5140, 5340, 5540, 5740, 5940, 6140, 6340, 6540, 6740, 6940, 7140, 7340, 7540, 7740, 7940, 8140, 8340, 8540, 8740, 8940, 9140, 9340, 9540, 9740, 9940, 10140, 10340, 10540, 10740, 10940, 11140, 11340, 11540, 11740, 11940, 12140, 12340, 12540, 12740, 12940, 13140, 13340, 13540, 13740, 13940, 14140, 14340, 14540, 14740, 14940, 15140, 15340, 15540, 15740, 15940, 16140, 16340, 16540, 16740, 16940, 17140, 17340, 17540, 17740, 17940, 18140, 18340, 18540, 18740, 18940, 19140, 19340, 19540, 19740, 19940, 20140, 20340, 20540, 20740, 20940, 21140, 21340, 21540, 21740, 21940, 22140, 22340, 22540, 22740, 22940, 23140, 23340, 23540, 23740, 23940, 24140, 24340, 24540, 2

In [159]:

recur1=patient['Recurred']
recur2=patient.Recurred
print("Recurred Column is:\n",recur1)
print()
print()
print()
print("Recurred Column is:\n",recur2)


Recurred Column is:
 patient_ID
2340      No
2540      No
2740      No
2940      No
3140      No
        ... 
77940    Yes
78140    Yes
78340    Yes
78540    Yes
78740    Yes
Name: Recurred, Length: 383, dtype: object



Recurred Column is:
 patient_ID
2340      No
2540      No
2740      No
2940      No
3140      No
        ... 
77940    Yes
78140    Yes
78340    Yes
78540    Yes
78740    Yes
Name: Recurred, Length: 383, dtype: object


In [164]:

#How many patients have been recurred
no_of_recur1=patient[patient['Recurred']=='Yes']    #Using Filtering
no_of_recur2=len(no_of_recur1)
                               #counts.get('Yes', 0): This will return the count of 'Yes' from the value_counts() Series.
                               # If 'Yes' is not found in the data, it returns 0 instead of raising an error or returning None.
no_of_recur3=patient['Recurred'].value_counts()                         
no_of_recur4=patient.Recurred.value_counts().get('Yes',0)                         
print("Number of patients who have recurred: ",no_of_recur2)
print()
print()
print("Number of patients who have recurred: ",no_of_recur3)
print()
print()
print("Number of patients who have recurred: ",no_of_recur4)



Number of patients who have recurred:  108


Number of patients who have recurred:  Recurred
No     275
Yes    108
Name: count, dtype: int64


Number of patients who have recurred:  108


In [165]:


#Most frequent recurrence
mst_freq_rec1=patient['Recurred'].value_counts().idxmax()
mst_freq_rec2=patient['Recurred'].value_counts().head(1)
mst_freq_rec3=patient['Recurred'].value_counts().head(1).index[0]
print("Most frequent recurrence is: ",mst_freq_rec1)
print()
print()
print("Most frequent recurrence is: ",mst_freq_rec2)
print()
print()
print("Most frequent recurrence is: ",mst_freq_rec3)





Most frequent recurrence is:  No


Most frequent recurrence is:  Recurred
No    275
Name: count, dtype: int64


Most frequent recurrence is:  No


In [166]:


#Summarize the Data Frame
sum1=patient.describe()    #by default returns numerical columns
print("Summary of Dataframe is:\n",sum1)
print()
print()
sum2=patient.describe(include='all')  
print("Summary of all columns is:\n",sum2.to_string())
print(patient.isnull().sum())
print()
print()


#Summary of only recurrance column is 
sum3=patient['Recurred'].describe()    
print("Summary of Recurred column is:",sum3)   
print()
print()


#Mean age of the patients
mean_age=patient['Age'].mean()
print("Mean age of users is: ",mean_age)
print()
print()
#Age with least occurance 
age=patient['Age'].value_counts().tail(10)
print("Patients age with least occurance is:",age)

Summary of Dataframe is:
               Age
count  383.000000
mean    40.866841
std     15.134494
min     15.000000
25%     29.000000
50%     37.000000
75%     51.000000
max     82.000000


Summary of all columns is:
                Age Gender Smoking Hx Smoking Hx Radiothreapy Thyroid Function Physical Examination Adenopathy  Pathology   Focality Risk    T    N    M Stage   Response Recurred
count   383.000000    383     383        383             383              383                  383        383        383        383  383  383  383  383   383        383      383
unique         NaN      2       2          2               2                5                    5          6          4          2    3    7    3    2     5          4        2
top            NaN      F      No         No              No        Euthyroid  Multinodular goiter         No  Papillary  Uni-Focal  Low   T2   N0   M0     I  Excellent       No
freq           NaN    312     334        355             376          

- count: The number of non-missing (non-null) values in each column. For example, count for "Age" is 383, indicating that there are 383 non-null values in the "Age" column.

- unique: The number of unique values in each column. For categorical columns (like "Gender" and "Smoking"), this shows how many distinct categories are present.





- top: The most frequently occurring value in the column. For categorical columns, this shows the most common category. For example, the "Gender" column has 'F' as the most common value.

- freq: The frequency of the most common value (i.e., how many times the value in the "top" row appears). For example, 'F' appears 312 times in the "Gender" column.

- mean: The average value of the column, applicable to numerical columns. For example, the mean age is approximately 40.87. This is not applicable to categorical columns, so those will show NaN (Not a Number).

- std: The standard deviation, which measures the spread or dispersion of the numerical values in the column. For the "Age" column, the standard deviation is approximately 15.13. This is also not applicable to categorical columns.

- min: The minimum value in the column for numerical data. For the "Age" column, the minimum age is 15.

- 25%: The 25th percentile value, meaning that 25% of the data is less than or equal to this value. For the "Age" column, 25% of the ages are 29 or younger.

- 50%: The median value, meaning that 50% of the data is less than or equal to this value. For "Age," the median age is 37.

- 75%: The 75th percentile value, meaning that 75% of the data is less than or equal to this value. For "Age," 75% of the ages are 51 or younger.

- max: The maximum value in the column for numerical data. For the "Age" column, the maximum age is 82.

- For Categorical Columns:
- unique, top, and freq are relevant, showing the diversity and most common values.
- mean, std, min, 25%, 50%, 75%, and max are not applicable and will show NaN.

In [163]:

dataset='Thyroid_Diff.csv'
patient=pd.read_csv(dataset)
numeric_cols = patient.describe()
print(numeric_cols)
mean_dict = {}
mode_dict = {}
median_dict = {}
std_dev_dict = {}
variance_dict = {}
max_dict = {}
min_dict = {}


for col in numeric_cols:
    mean_dict[col] = patient[col].mean()
    #mean_dict['Age']=patient['Age'].mean()
    mode_dict[col] = patient[col].mode().values[0]
    #mode_dict['Age'] = patient['Age'].mode().values[0]
    #values[0] is used to extract the first value from the Series returned by .mode() 
    # to handle cases where multiple modes might be returned, ensuring you get a single value for simplicity.
    median_dict[col] = patient[col].median()
    #median_dict['Age'] = patient['Age'].median()
    std_dev_dict[col] = patient[col].std()
    #std_dev_dict['Age'] = patient['Age'].std()
    variance_dict[col] = patient[col].var()
    #variance_dict['Age'] = patient['Age'].var()
    max_dict[col] = patient[col].max()
    #max_dict['Age'] = patient['Age'].max()
    min_dict[col] = patient[col].min()
    #min_dict['Age'] = patient['Age'].min()
    


print("Mean:\n", mean_dict)
print("\nMode:\n", mode_dict)
print("\nMedian:\n", median_dict)
print("\nStandard Deviation:\n", std_dev_dict)
print("\nVariance:\n", variance_dict)
print("\nMaximum:\n", max_dict)
print("\nMinimum:\n", min_dict)    
 
 
    


              Age
count  383.000000
mean    40.866841
std     15.134494
min     15.000000
25%     29.000000
50%     37.000000
75%     51.000000
max     82.000000
Mean:
 {'Age': np.float64(40.866840731070496)}

Mode:
 {'Age': np.int64(31)}

Median:
 {'Age': np.float64(37.0)}

Standard Deviation:
 {'Age': np.float64(15.134493807858814)}

Variance:
 {'Age': np.float64(229.05290282011677)}

Maximum:
 {'Age': np.int64(82)}

Minimum:
 {'Age': np.int64(15)}
